src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,760 @@
+import { default as flags } from 'workerd:compatibility-flags';
+
+export const kConnectionsCheckingInterval = Symbol(
+  'http.server.connectionsCheckingInterval'
+);
+
+let serverRegistry: FinalizationRegistry<number> | null = null;
+
+// Since finalization registry is only available under a specific compat flag,
+// let's check if it's enabled to preserve backward compatibility.
+if (flags.jsWeakRef) {
+  serverRegistry = new FinalizationRegistry((port) => {
+    portMapper.delete(port);
+  });
+}
+
+export type StreamController = ReadableStreamDefaultController<Uint8Array>;
+
+export type DataWrittenEvent = {
+  index: number;
+  entry: WrittenDataBufferEntry;
+};
+
+export class Server
+  extends EventEmitter
+  implements _Server, BaseWithHttpOptions
+{
+  // @ts-expect-error TS2416 Server is not assignable to same property in base type.
+  [kIncomingMessage]: typeof IncomingMessage = IncomingMessage;
+  // @ts-expect-error TS2416 Server is not assignable to same property in base type.
+  [kServerResponse]: typeof ServerResponse = ServerResponse;
+  [kConnectionsCheckingInterval]?: number;
+  [kUniqueHeaders]: Set<string> | null = null;
+
+  // Similar option to this. Too lazy to write my own docs.
+  // http://www.squid-cache.org/Doc/config/half_closed_clients/
+  // https://wiki.squid-cache.org/SquidFaq/InnerWorkings#What_is_a_half-closed_filedescriptor.3F
+  httpAllowHalfOpen = false;
+  timeout = 0;
+  maxHeadersCount: number | null = null;
+  maxRequestsPerSocket = 0;
+  connectionsCheckingInterval = 30_000;
+  requestTimeout: number = 0;
+  headersTimeout: number = 0;
+  requireHostHeader: boolean = false;
+  joinDuplicateHeaders: boolean = false;
+  rejectNonStandardBodyWrites: boolean = false;
+  keepAliveTimeout: number = 5_000;
+  port: number | null = null;
+
+  constructor(options?: ServerOptions, requestListener?: RequestListener) {
+    if (!flags.enableNodejsHttpServerModules) {
+      throw new ERR_METHOD_NOT_IMPLEMENTED('Server');
+    }
+    super();
+
+    if (options != null) {
+      // @ts-expect-error TS2345 TODO(soon): Find a better way to handle this type mismatch.
+      storeHTTPOptions.call(this, options);
+    }
+
+    // TODO(soon): Support options.highWaterMark option.
+
+    if (typeof options === 'function') {
+      requestListener = options;
+      options = {};
+    } else if (options == null) {
+      options = {};
+    } else {
+      validateObject(options, 'options');
+    }
+
+    if (requestListener) {
+      this.on('request', requestListener);
+    }
+
+    this[kUniqueHeaders] = parseUniqueHeadersOption(
+      options.uniqueHeaders as (string | string[])[]
+    );
+  }
+
+  close(callback?: VoidFunction): this {
+    httpServerPreClose(this);
+    if (this.port != null) {
+      portMapper.delete(this.port);
+      serverRegistry?.unregister(this);
+      this.port = null;
+    }
+    this.emit('close');
+    queueMicrotask(() => {
+      callback?.();
+    });
```

Comment:
The `callback` passed into `close` is registered as a `'close'` event handler and the `close` event should be emitted in the `queueMicrotask(...)`

```suggestion
    if (typeof callback === 'function') {
      this.on('close', callback);
    }
    queueMicrotask(() => {
      this.emit('close');
    });
```

For instance, in Node.js:

```bash
jsnell@james-cloudflare-build:~/tmp$ node a
after close
closed 1
closed 2
jsnell@james-cloudflare-build:~/tmp$ cat a.js
const { createServer } = require('node:http');

const server = createServer();

server.on('close', () => console.log('closed 1'));

server.listen(8888);

server.close(() => console.log('closed 2'));

console.log('after close');
jsnell@james-cloudflare-build:~/tmp$ 
```

Reply:
Not sure why this was marked resolved as it appears to still be an issue.